File: app/src/main/java/com/trevorwiebe/trackacow/data/local/dao/DrugsGivenDao.java
package com.trevorwiebe.trackacow.data.local.dao;

import androidx.room.Dao;
import androidx.room.Delete;
import androidx.room.Insert;
import androidx.room.OnConflictStrategy;
import androidx.room.Query;
import androidx.room.Update;

import com.trevorwiebe.trackacow.data.local.entities.DrugsGivenEntity;

import java.util.List;

@Dao
public interface DrugsGivenDao {

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    void insertDrugsGiven(DrugsGivenEntity drugsGivenEntity);

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    void insertDrugsGivenList(List<DrugsGivenEntity> drugsGivenEntities);

    @Query("SELECT * FROM DrugsGiven WHERE cowId = :cowId")
    List<DrugsGivenEntity> getDrugsGivenByCowId(String cowId);

    @Query("SELECT * FROM DrugsGiven WHERE cowId IN(:cowIdList)")
    List<DrugsGivenEntity> getDrugsGivenByCowIdList(List<String> cowIdList);

    @Query("SELECT * FROM DrugsGiven WHERE lotId IN(:lotIds)")
    List<DrugsGivenEntity> getDrugsGivenByLotIds(List<String> lotIds);

    @Query("SELECT * FROM DrugsGiven WHERE drugGivenId = :drugGivenId")
    DrugsGivenEntity getDrugGivenByDrugGivenId(String drugGivenId);

    @Query("SELECT * FROM DrugsGiven WHERE (lotId = :lotId) AND (date BETWEEN :startDate AND :endDate)")
    List<DrugsGivenEntity> getDrugsGivenByLotIdAndDateRange(String lotId, long startDate, long endDate);

    @Query("UPDATE DrugsGiven SET amountGiven = :amountGiven WHERE drugGivenId = :drugGivenId")
    void updateDrugGivenAmountGiven(int amountGiven, String drugGivenId);

    @Query("UPDATE DrugsGiven Set date = :date WHERE cowId = :cowId")
    void updateDrugsGivenDateByCowId(long date, String cowId);

    @Query("DELETE FROM DrugsGiven WHERE lotId = :lotId")
    void deleteDrugsGivenByLotId(String lotId);

    @Query("DELETE FROM DrugsGiven WHERE cowId = :cowId")
    void deleteDrugsGivenByCowId(String cowId);

    @Query("DELETE FROM DrugsGiven WHERE drugGivenId = :drugId")
    void deleteDrugGivenById(String drugId);

    @Query("DELETE FROM DrugsGiven")
    void deleteDrugsGivenTable();

    @Update
    void updateDrugsGiven(DrugsGivenEntity drugsGivenEntity);

    @Delete
    void deleteDrugsGiven(DrugsGivenEntity drugsGivenEntity);

}

File: app/src/main/java/com/trevorwiebe/trackacow/data/local/dao/DrugsGivenDao.kt
package com.trevorwiebe.trackacow.data.local.dao

import androidx.room.*
import com.trevorwiebe.trackacow.data.entities.DrugsGivenEntity
import com.trevorwiebe.trackacow.data.entities.compound_entities.DrugsGivenAndDrugEntity
import kotlinx.coroutines.flow.Flow

@Dao
interface DrugsGivenDao {


    @Insert(onConflict = OnConflictStrategy.IGNORE)
    suspend fun insertDrugsGivenList(drugsGivenList: List<DrugsGivenEntity>): List<Long>

    @Query(
            "SELECT * FROM drugs_given " +
                    "INNER JOIN drug ON drug.drugCloudDatabaseId = drugs_given.drugsGivenDrugId " +
                    "WHERE drugsGivenLotId = :lotId"
    )
    fun getDrugsGivenAndDrugByLotId(lotId: String): Flow<List<DrugsGivenAndDrugEntity>>

    @Query(
            "SELECT * FROM drugs_given " +
                    "INNER JOIN drug ON drug.drugCloudDatabaseId = drugs_given.drugsGivenDrugId " +
                    "WHERE drugsGivenLotId = :lotId AND drugsGivenDate BETWEEN :startDate AND :endDate"
    )
    fun getDrugsGivenAndDrugByLotIdAndDate(lotId: String, startDate: Long, endDate: Long):
            Flow<List<DrugsGivenAndDrugEntity>>

    @Query(
            "SELECT * FROM drugs_given " +
                    "INNER JOIN drug ON drug.drugCloudDatabaseId = drugs_given.drugsGivenDrugId " +
                    "WHERE drugsGivenCowId IN (:cowIdList)"
    )
    fun getDrugsGivenAndDrugByCowId(cowIdList: List<String>): Flow<List<DrugsGivenAndDrugEntity>>

    @Query("SELECT * FROM drugs_given WHERE drugsGivenCowId = :cowId")
    suspend fun getDrugsGivenByCowId(cowId: String): List<DrugsGivenEntity>

    @Query("DELETE FROM drugs_given WHERE drugsGivenCowId = :cowId")
    suspend fun deleteDrugsGivenByCowId(cowId: String)

    @Update
    suspend fun updateDrugGiven(drugsGivenEntity: DrugsGivenEntity)

    @Query("UPDATE drugs_given SET drugsGivenLotId =:lotId WHERE drugsGivenLotId IN (:oldLotIds)")
    suspend fun updateDrugsGivenWithNewLotId(lotId: String, oldLotIds: List<String>)

    @Update
    suspend fun updateDrugsGivenList(drugsGivenList: List<DrugsGivenEntity>)

    @Delete
    suspend fun deleteDrugGiven(drugsGivenEntity: DrugsGivenEntity)

    @Query("DELETE FROM drugs_given")
    suspend fun deleteAllDrugsGiven()

    @Query("DELETE FROM drugs_given WHERE drugsGivenLotId = :lotId")
    suspend fun deleteDrugsGivenByLotId(lotId: String)

    @Query("DELETE FROM drugs_given WHERE drugsGivenLotId = :lotId AND drugsGivenDate BETWEEN :startDate AND :endDate")
    suspend fun deleteDrugsGivenByLotIdAndDate(lotId: String, startDate: Long, endDate: Long)

    @Transaction
    suspend fun deleteDrugsGivenByCowIdTransaction(cowId: String): List<DrugsGivenEntity> {
        val drugsGivenList = getDrugsGivenByCowId(cowId)
        deleteDrugsGivenByCowId(cowId)
        return drugsGivenList
    }

    @Transaction
    suspend fun syncCloudDrugsGivenByCowId(drugsGivenList: List<DrugsGivenEntity>, cowId: String) {
        deleteDrugsGivenByCowId(cowId)
        insertDrugsGivenList(drugsGivenList)
    }

    @Transaction
    suspend fun syncCloudDrugsGivenByLotId(drugsGivenList: List<DrugsGivenEntity>, lotId: String) {
        deleteDrugsGivenByLotId(lotId)
        insertDrugsGivenList(drugsGivenList)
    }

    @Transaction
    suspend fun syncCloudDrugsGivenByLotIdAndDate(
        drugsGivenList: List<DrugsGivenEntity>,
        lotId: String,
        startDate: Long,
        endDate: Long
    ) {
        deleteDrugsGivenByLotIdAndDate(lotId, startDate, endDate)
        insertDrugsGivenList(drugsGivenList)
    }

}